src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java

```diff
@@ -355,58 +349,88 @@ public void handleEvent(Event event) {
-            if (isPhyssimMode(mode)) {
-                traversalEventsForPhysSimulation.add((PathTraversalEvent) event);
-
-                double departureTime = pte.departureTime();
-                String vehicleId = pte.vehicleId().toString();
-
-                String vehicleType = pte.vehicleType();
-                Id<BeamVehicleType> beamVehicleTypeId = Id.create(vehicleType, BeamVehicleType.class);
-                boolean isCaccEnabled = beamServices.beamScenario().vehicleTypes().get(beamVehicleTypeId).get().isCaccEnabled();
-                caccVehiclesMap.put(vehicleId, isCaccEnabled);
-
-                // For every PathTraversalEvent which has PhysSim mode (CAR or BUS) we create
-                // - If person does not exist, we create Person from `vehicleId`. For that person we create plan, set it to selected plan and add attributes from the original person
-                // - Create leg
-                // - Create dummy activity
-                final Person person = initializePersonAndPlanIfNeeded(Id.createPersonId(vehicleId), Id.createPersonId(pte.driverId()));
-                final Plan plan = person.getSelectedPlan();
-                final Leg leg = createLeg(pte);
-
-                if (leg == null) {
-                    return; // dont't process leg further, if empty
-                }
+        }
+        // pt sampling
+        // TODO: if requested, add beam.physsim.ptSamplingMode (pathTraversal | busLine), which controls if instead of filtering outWriter
+        // pathTraversal, a busLine should be filtered out, avoiding jumping buses in visualization (but making traffic flows less precise).
 
-                Activity previousActivity = jdeqsimPopulation.getFactory().createActivityFromLinkId(DUMMY_ACTIVITY, leg.getRoute().getStartLinkId());
-                previousActivity.setEndTime(departureTime);
-                plan.addActivity(previousActivity);
-                plan.addLeg(leg);
+        if (mode.equalsIgnoreCase(BUS) && rand.nextDouble() > beamConfig.beam().physsim().ptSampleSize()) {
+            return;
+        }
+        if (isPhyssimMode(mode)) {
+            traversalEventsForPhysSimulation.add(pte);
+
+            String driverId = pte.driverId();
+            String vehicleId = pte.vehicleId().toString();
+
+            String vehicleType = pte.vehicleType();
+            Id<BeamVehicleType> beamVehicleTypeId = Id.create(vehicleType, BeamVehicleType.class);
+            boolean isCaccEnabled = beamServices.beamScenario().vehicleTypes().get(beamVehicleTypeId).get().isCaccEnabled();
+            caccVehiclesMap.put(vehicleId, isCaccEnabled);
+
+            addPTEtoPhysSimPlans(pte, vehicleId, driverId, 0);
+
+            double fractionOfEvents = beamConfig.beam().physsim().duplicatePTE().fractionOfEventsToDuplicate();
+            int iterationOfAdding = 0;
+            if (mode.equalsIgnoreCase(CAR)) {
+                while (rand.nextDouble() < (fractionOfEvents - iterationOfAdding)) {
```

Comment:
I didn't quite get what this loop is doing. rand.nextDouble returns a double in a range [0,1}. That most of the time is lower than the difference on the right. Is this something similar to `beam.utils.MathUtils#roundUniformly` this method?